### PR TITLE
Feature/sincal access db reader

### PR DESCRIPTION
**Pull request by SamRWest**
Added Sincal reader for sqllite and access databases and improved it to support sqllite & access database formats via sqlalchemy.  
Added sincal test data
Added network visualisation functions.

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (thomasbrinsmead) [inline .github/workflows/ci.yml:1]: Great use of shading to highlight deletions and insertions!

File: ditto/readers/abstract_lv_reader.py
Comment (thomasbrinsmead) [inline ditto/readers/abstract_lv_reader.py:1]: Line numbering is consecutive, except for a few missing, like between 18/19, 48/49, 88/89. Not all lines that end in "8" though.

File: ditto/readers/sincal/__init__.py
Comment (thomasbrinsmead) [inline ditto/readers/sincal/__init__.py:1]: importing from the future appears to violate causality. how do you propose to resolve the grandfather paradox?

File: ditto/readers/sincal/exception_logger.py
Comment (thomasbrinsmead) [inline ditto/readers/sincal/exception_logger.py:1]: Great alignment of plus symbols.

File: ditto/readers/sincal/read.py
Comment (thomasbrinsmead) [inline ditto/readers/sincal/read.py:1]: What does CLI stand for?

Comment (thomasbrinsmead) [reply ditto/readers/sincal/read.py:1]: Are repeated if assignments suited to use of a dictionary structure? (230ff)

Comment (thomasbrinsmead) [reply ditto/readers/sincal/read.py:1]: This sounds like a stupid question, but it is serious. Given the large number of methods for this class, doesn't it make sense to give it some kind of additional structure? Like maybe an upfront index with method categories? 

There are a few database connection methods, followed by methods for reading individual network elements, and then some multithreading methods and then what looks like a key "main" method to parse the whole network.

And the network elements are categorised into conductor type elements and node type elements, generation elements, transformer elements, etc. But a lot of this structure is highly implicit. Is is possible (or recommended by coding practice standards) to explicitly provide a view of the interdependency structure of the methods?

Some note about how the main method is intentionally designed in a somewhat recursive manner, so to understand how this class should mostly be used, apart from the init method, you should focus on the "parse" or "parse_network" method, and observe that it is built up from reading individual elements, using class methods. What should read_multi_threaded be considered an alternative to?

File: ditto/metrics/network_analysis.py
Comment (thomasbrinsmead) [inline ditto/metrics/network_analysis.py:1]: Seems harmless.

File: ditto/readers/sincal/read_capacitors.py
Comment (thomasbrinsmead) [inline ditto/readers/sincal/read_capacitors.py:1]: Could do with more comments.

File: ditto/readers/sincal/read_lines.py
Comment (thomasbrinsmead) [inline ditto/readers/sincal/read_lines.py:1]: Lines 168ff of abstract_lv_reader and lines 47-57 of read_lines are making use of essentially the identical (Clark transform?) matrix, defined in slightly different ways. Would it make sense to define this matrix once as a constant for the class and use it within each method? Or have a single method that defines the matrix that is called by each method?

Comment (thomasbrinsmead) [reply ditto/readers/sincal/read_lines.py:1]: The parse_line method is very large, and hard to read by a human. It doesn't help that there is a large quantity of commented out code within it. Are there good coding conventions for keeping this history (and enabling resurrecting previous experimental code) while still making it easy to read the working code?

When I talk about wanting tools and/or conventions for helping me manage code development, I am talking about or imagining methods to help me to document the experiment that I did to decide to leave THAT bit of code disabled, and THAT version of the CODE enabled, why did I leave that bit of code commented out instead of deleted, even how many times and under what circumstances unit tests or deployments of THAT version of the method have been executed. 

And these sort of coding decisions take place at several scales, so it's hard to keep track of it all in my head.

File: ditto/readers/sincal/read_lines_v2.py
Comment (thomasbrinsmead) [inline ditto/readers/sincal/read_lines_v2.py:1]: Here is a good case in point related to my previous comment. This is apparently a "version 2" alternative of a method. How do they differ? Why are they both included? Is this supposed to be experimental, or are they to be deployed in different circumstances? What tests have already been done (if any) to compare their performance? What tradeoffs is the programmer expecting to see? Is one going to be deprecated? Is the difference as big as a version 1 versus 2 for a project? Or is it just 1.0 versus 1.1, or 1.3.2.5 versus 1.3.2.6?

Comment (thomasbrinsmead) [reply ditto/readers/sincal/read_lines_v2.py:1]: Obviously there are a lot of common aspects of version 2 and the original of this method. At what point in the process do we write that code once and reuse it versus leaving in copies?

File: ditto/readers/sincal/read_loads.py
Comment (thomasbrinsmead) [inline ditto/readers/sincal/read_loads.py:1]: More silly questions like - is there are standard way that I would find out whether and how often and/or how this method has passed a unit test? Or whether there is a unit test for it? How big is a unit to be tested?

File: ditto/readers/sincal/read_nodes.py
Comment (thomasbrinsmead) [inline ditto/readers/sincal/read_nodes.py:1]: I remember writing some absolutely awful hacks to some of these methods, having no idea what I was doing software-wise. Quite likely some of these hacks have been reversed, not included, ignored, or maybe they haven't been seen at all by you or others. But some of them did kind of solve power systems issues. Even if the solutions that were implemented were inelegant or even awful, what would be an appropriate way of revisiting some of them to at least identify bugs that they were trying to solve? (I'm not sure it would be worth the effort, but what would the effort look like?)

File: ditto/readers/sincal/read_reactors.py
Comment (thomasbrinsmead) [inline ditto/readers/sincal/read_reactors.py:1]: Again there is lots of code that is repeated between methods. Should we take advantage of that to simplify the totality of the code? Is that what would be done during a "refactor"?

File: ditto/readers/sincal/read_sources.py
Comment (thomasbrinsmead) [inline ditto/readers/sincal/read_sources.py:1]: How could I have been informed that this code wasn't written from scratch, but is actually copied from elsewhere? Should we record the legacy of where it comes from, just in case the "elsewhere" is subsequently upgraded? How well tested was it already when it was elsewhere? Were there unit tests elsewhere as well?

File: ditto/visualisation/__init__.py
Comment (thomasbrinsmead) [inline ditto/visualisation/__init__.py:1]: Why?

File: ditto/readers/sincal/read_transformers.py
Comment (thomasbrinsmead) [inline ditto/readers/sincal/read_transformers.py:1]: So much apparently repeated code. Could become a bit of nightmare when maintaining. When you modify one instance of the copy, should you copy the same modification everywhere else? Or do you just write the code snippet once and modify the single canonical instance? Now you might have to test every place that the instance is used?

File: ditto/visualisation/vis_utils.py
Comment (thomasbrinsmead) [inline ditto/visualisation/vis_utils.py:1]: This is a nice shade of green.

I see that there a rudimentary visualisation method defined here. Has this been made completely redundant now by work done by Bis? Or should it be kept because it saves having to import the other library? If we were to keep both, should some of these be designed to have the same interfaces for the purposes of direct substitution? Do these principles hold if the languages were different?

File: ditto/writers/opendss/write.py
Comment (thomasbrinsmead) [inline ditto/writers/opendss/write.py:1]: Gee.

File: pyproject.toml [hunk 1]
```diff
@@ -36,6 +36,13 @@ sphinxcontrib-napoleon = { optional = true, version = "*" }
 xlrd = { optional = true, version = "*" }
 xlsxwriter = { optional = true, version = "*" }
 
+sqlalchemy = { version = "^2.0.16", optional = true }
+sqlalchemy-access = { version = "^2.0.2", optional = true, platform = "win32" }
+
+matplotlib = { version = ">=3.7.1", optional = true }
+pyvis = { version = "^0.1.8", optional = true }
+plotly = { version = ">=5.15.0", optional = true }
+
```
Comment (thomasbrinsmead) [inline pyproject.toml:45]: Look, I can leave comments on specific rows!

File: tests/data/big_cases/sincal/LVFT-Network-K/database.db
Comment (thomasbrinsmead) [inline tests/data/big_cases/sincal/LVFT-Network-K/database.db:1]: Binary file not inspected.

File: tests/data/small_cases/sincal/NFTS_Representative_19/database.mdb
Comment (thomasbrinsmead) [inline tests/data/small_cases/sincal/NFTS_Representative_19/database.mdb:1]: Ditto.

File: tests/readers/sincal/__init__.py
Comment (thomasbrinsmead) [inline tests/readers/sincal/__init__.py:1]: Another empty file.

File: tests/readers/sincal/test_sincal_reader.py
Comment (thomasbrinsmead) [inline tests/readers/sincal/test_sincal_reader.py:1]: A unit test. Good.

File: poetry.lock
Comment (thomasbrinsmead) [inline poetry.lock:1]: This file is automatically generated by poetry and should not be changed by hand.

File: ditto/readers/sincal/read_photovoltaics.py [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+from ditto.models.phase_winding import PhaseWinding
+from ditto.models.regulator import Regulator
+from ditto.models.wire import Wire
+from ditto.models.capacitor import Capacitor
+from ditto.models.phase_capacitor import PhaseCapacitor
+from ditto.models.reactor import Reactor
+from ditto.models.phase_reactor import PhaseReactor
+from ditto.models.photovoltaic import Photovoltaic
+from ditto.readers.sincal.exception_logger import log_exceptions
+
```
Comment (thomasbrinsmead) [inline ditto/readers/sincal/read_photovoltaics.py:36]: There are two methods. One parses a single element. The other parses numerous instances of that element, calling the single element method iteratively.

**Review by thomasbrinsmead — APPROVED** [review]
I took longer than recommended. But it is my first time doing a code review, so I'll some of that time to training.

File: .github/workflows/ci.yml
Comment (SamRWest) [reply .github/workflows/ci.yml:1]: Haha, I'll assume that was a joke :)

File: ditto/metrics/network_analysis.py
Comment (SamRWest) [reply ditto/metrics/network_analysis.py:1]: Yeah, this was actually a fix for the original ditto source code - was (quite rightly) throwing errors in python3.7

File: ditto/readers/abstract_lv_reader.py
Comment (SamRWest) [reply ditto/readers/abstract_lv_reader.py:1]: Do you mean the line numbering?  
(Github only shows lines around the changes lines)

File: ditto/readers/sincal/__init__.py
Comment (SamRWest) [reply ditto/readers/sincal/__init__.py:1]: by invoking the alternative universes time-travel paradigm.

File: ditto/readers/sincal/exception_logger.py
Comment (SamRWest) [reply ditto/readers/sincal/exception_logger.py:1]: umm, thanks?

File: ditto/readers/sincal/read.py
Comment (SamRWest) [reply ditto/readers/sincal/read.py:1]: CLI = Command Line Interface

> Are repeated if assignments suited to use of a dictionary structure? (230ff)

Do you mean the massive chunk of  ` if vectorGroup == 1 and winding == 0: connection = "D"` blocks?
Then yeah, this is horrible.  But (I should have mentioned) that I've intentionally not changed any of the in-depth parser code in this PR, as I intended _just_ to get it integrated, working and released.  But yes, packing this into a DataFrame indexed by vectorGroup and winding would be a much nicer solution here.

There's definitely a lot of room for improvement in the code and doc (maybe in a future PR), but if it ain't broke I'm ignoring it for now.

> Is is possible (or recommended by coding practice standards) to explicitly provide a view of the interdependency structure of the methods?

Yeah, some doc on how the sincal format and parsers are structured would be useful.  Despite getting this integrated, I'm still not sure about this myself though - there's a lot of detail in there to sift through...

File: ditto/readers/sincal/read_capacitors.py
Comment (SamRWest) [reply ditto/readers/sincal/read_capacitors.py:1]: Agreed.  I'd love to, but have no idea how this works TBH.

File: ditto/readers/sincal/read_lines.py
Comment (SamRWest) [reply ditto/readers/sincal/read_lines.py:1]: > Lines 168ff of abstract_lv_reader and lines 47-57 of read_lines are making use of essentially the identical (Clark transform?) matrix, defined in slightly different ways. Would it make sense to define this matrix once as a constant for the class and use it within each method? Or have a single method that defines the matrix that is called by each method?

Good catch, I didn't realise that. Would have to spend some time testing to make sure these methods are identical, but yes it'd be good to remove the duplication, and replace the read_lines version with the function if possible.  TBD.

Comment (SamRWest) [reply ditto/readers/sincal/read_lines.py:1]: > The parse_line method is very large, and hard to read by a human. It doesn't help that there is a large quantity of commented out code within it. Are there good coding conventions for keeping this history (and enabling resurrecting previous experimental code) while still making it easy to read the working code?
> 
> When I talk about wanting tools and/or conventions for helping me manage code development, I am talking about or imagining methods to help me to document the experiment that I did to decide to leave THAT bit of code disabled, and THAT version of the CODE enabled, why did I leave that bit of code commented out instead of deleted, even how many times and under what circumstances unit tests or deployments of THAT version of the method have been executed.
> 
> And these sort of coding decisions take place at several scales, so it's hard to keep track of it all in my head.

I totally agree that these methods are hard to follow, and the big commented block needs to be removed or explained.  As I have little idea how these work in practice though, I'm hesitant to remove them, so there are clues for the next person trying to debug this code.

I general, we shouldn't leave big commented blocks of code like this in here.  Experiments should be performed on separate branches and only merged when clean and complete, and accompanied by unit tests, which often explain the decisions made by example.

File: ditto/readers/sincal/read_lines_v2.py
Comment (SamRWest) [reply ditto/readers/sincal/read_lines_v2.py:1]: Good question.  I had a closer look, and it appears that Eric copies read_lines and modified it to merge consecutive lines with identical characteristics into a single Line object (for models derived from GIS data with poles as Nodes etc).  This could certainly be improved if someone had time/motivation. I've added some doc and renamed the class to make this a bit more obvious.

File: ditto/readers/sincal/read_loads.py
Comment (SamRWest) [reply ditto/readers/sincal/read_loads.py:1]: Yep, sure is.  
You can see individual test coverage in the Actions tab (eg [here](https://github.com/csiro-energy-systems/ditto/actions/runs/5384988232/jobs/9773528526#step:6:17))
And you can run `pytest` and look at the resulting HTML coverage reports for individual line coverage.  If a line is 'covered', it means it was run by on or more passing unit tests.

File: ditto/readers/sincal/read_nodes.py
Comment (SamRWest) [reply ditto/readers/sincal/read_nodes.py:1]: If they solved a real problem, and aren't causing speed or parsing issues (which they're not afaik), I'd probably leave them alone. The code doesn't look horrible, so I'd be interested to hear about what the hacks were...

File: ditto/readers/sincal/read_reactors.py
Comment (SamRWest) [reply ditto/readers/sincal/read_reactors.py:1]: Yep, that's a refactor.  Easy at the time of writing, hard after the fact when the author isn't around any more.

File: ditto/readers/sincal/read_sources.py
Comment (SamRWest) [reply ditto/readers/sincal/read_sources.py:1]: The full git history is still in our internal 'ditto-csiro' bitbucket repo, so you'd look there.
I chose to copy the parser into a fresh fork instead of maintaining this history to make a potential PR back into NREL's repo more self-contained and concise.
And no, sadly there were zero tests in the original fork, apart from those I added during LVFT.

File: ditto/readers/sincal/read_transformers.py
Comment (SamRWest) [reply ditto/readers/sincal/read_transformers.py:1]: Yeah, it's pretty painful to maintain - lots of replace-all and run test to check.
Turning the repeated code into re-used functions would be good, given time.

File: ditto/visualisation/__init__.py
Comment (SamRWest) [reply ditto/visualisation/__init__.py:1]: __init__.py is a just standard way of marking a subdir as a python module.

File: ditto/visualisation/vis_utils.py
Comment (SamRWest) [reply ditto/visualisation/vis_utils.py:1]: This is the code used for all the LVFT visualisations.  Definitely worth keeping so we can render future networks without bus coordinates using the force-directed graph stuff.  

Bis's notebooks are great, but use code from dss-python (and require bus coordinates), so are also harder to modify and maintain and aren't interactive.  These functions also work with any format Ditto can read, not just DSS.

Oh, and the library imports are optional, so aren't installed unless you do `pip install ditto[vis] | poetry add --extras vis`

File: ditto/writers/opendss/write.py
Comment (SamRWest) [reply ditto/writers/opendss/write.py:1]: Yeah, not sure how that one made it through the tests

File: poetry.lock
Comment (SamRWest) [reply poetry.lock:1]: Yup

File: pyproject.toml (hunk 1 above)
Comment (SamRWest) [reply pyproject.toml:45]: Great!  That'll make it much easier to work out what you're actually talking about :p

**Review by SamRWest — COMMENTED** [review]
Thanks very much for the review Thomas.  I've replied to and/or resolved all your comments now, and pushed another commit fixing some of the issues.
